### PR TITLE
Goci 2460 ds generating gene documents

### PR DESCRIPTION
**Pull request by DSuveges**
A large set of code enabling the generation of gene documents for the slim solr index. Includes the addition of three classes: 
**REST** - a wrapper to fetch data from Ensembl's REST API
**gene_sql** - a class to extract gene related data from the database (Ensembl ID, rsIDs, association and study counts)
**GeneAnnotator** - tool to integrate gene annotation from various sources including Ensembl's REST API, HGNC, Ensembl's ftp directories.

Source files for Ensembl and HGNC are read from exported shell variables, so the wrapper (start.sh) was also modified.

### CODE REVIEW

File: scripts/document_types/gene_annotator.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        '''
+        Extract HGNC data from /nfs. Extracting relevant columns and formats
+
+        1. Select certain fields: alternative names/synonyms + cros ref IDs
+        2. Pool these names together + ensembl ID and entrez ID are separate columns.
+        3. Adding dataframe to object.
+        '''
+
+        # The file name needs to be read from the environment:
+        HGNC_file = '/nfs/ftp/pub/databases/genenames/new/tsv/non_alt_loci_set.txt'
```
Comment (twhetzel) [inline scripts/document_types/gene_annotator.py]: This filepath reference should be removed.

File: solr_config/schema.xml [hunk 2]
```diff
@@ -180,18 +179,16 @@
...
+  <field name="cytobands" type="text_general" indexed="true" stored="true" multiValued="false" />  
   <field name="biotype" type="text_general" indexed="true" stored="true" multiValued="false" />
-  <field name="variants" type="string" indexed="true" stored="true" multiValued="true" />
-
-  <field name="mapping" type="text_general" indexed="true" stored="true" multiValued="true" />  
-  <field name="mapping.chr" type="text_general" indexed="true" stored="true" multiValued="false" />  
-  <field name="mapping.start" type="int" indexed="true" stored="true" multiValued="false" />  
-  <field name="mapping.end" type="int" indexed="true" stored="true" multiValued="false" />  
+  <field name="rsIDs" type="string" indexed="true" stored="true" multiValued="true" />
+  <field name="synonymsGene" type="text_general" indexed="true" stored="true" multiValued="false" />
```
Comment (twhetzel) [inline solr_config/schema.xml]: Are "synonymsGene" intended to be searchable and displayed on the site? Currently gene synonyms can not be used in search and are not displayed on the site. I'd suggest removing this from the `schema.xml` document AND also the script that collects these.

File: solr_config/schema.xml [hunk 3]
```diff
@@ -180,18 +179,16 @@
...
+  <field name="crossRefs" type="text_general" indexed="true" stored="true" multiValued="false" />
```
Comment (twhetzel) [inline solr_config/schema.xml]: Is the expectation that users would search on "crossRefs"? If no, then this indexed can be set to false.

File: scripts/EnsemblREST/REST.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+from tqdm import tqdm
+
+class REST(object):
+    '''
+    This class will be responsible to return data from Ensembl.
+    It will be initialized by the URL of the server. So in theory it is easy to
+    switch to alternative or to the GRCh37 server.
+    '''
+
+    def __init__(self, URL = 'https://rest.ensembl.org'):
```
Comment (twhetzel) [inline scripts/EnsemblREST/REST.py]: The `generate_solr_docs.py` script uses the filename "https://jul2018.rest.ensembl.org/" on Line 317 as the default due to issues with the most recent Ensembl file, should the same link be used here too?

File: scripts/generate_solr_docs.py [hunk 5]
```diff
@@ -91,6 +90,38 @@ def save_data(data, data_type=None):
 def variant_data(connection, limit=0):
     return variant.get_variant_data(connection, limit)
 
+def gene_data(connection, limit=0):
```
Comment (twhetzel) [inline scripts/generate_solr_docs.py]: Following the existing code design, the details of how to create the document_type _should_ be in a separate script to leave this one more general.

File: scripts/document_types/gene_annotator.py (hunk 1 above)
Comment (DSuveges) [reply scripts/document_types/gene_annotator.py]: It's removed.

File: solr_config/schema.xml (hunk 2 above)
Comment (DSuveges) [reply solr_config/schema.xml]: I don't want to remove that at this point. I'll experiment with different setting to make sure we are still returning gene for their synonyms wile not polluting the search results.

File: solr_config/schema.xml (hunk 3 above)
Comment (DSuveges) [reply solr_config/schema.xml]: It's sometimes happen that users search for Entrez IDs or refseq IDs.

File: scripts/EnsemblREST/REST.py (hunk 4 above)
Comment (DSuveges) [reply scripts/EnsemblREST/REST.py]: I removed the default REST URL from everywhere. The REST URL is set in the `generate_solr_docs.py` which needs to updated once we are moving to e95

File: solr_config/schema.xml (hunk 2 above)
Comment (twhetzel) [reply solr_config/schema.xml]: Is this additional work requested or an experiementation?

File: solr_config/schema.xml (hunk 3 above)
Comment (twhetzel) [reply solr_config/schema.xml]: Ok, I thought in another conversation it was mentioned this doesn't occur, which is why I was asking. Looks good then!

File: scripts/generate_solr_docs.py (hunk 5 above)
Comment (DSuveges) [reply scripts/generate_solr_docs.py]: Addressed. Now all gene related code was moved to `gene.py`.

**Comment by DSuveges** [comment]
* Gene synonyms removed from the gene documents.
* Gene synonyms annotation removed from the solr shema.
* The address of the REST API is specified only once, pointing to the right source.
* Internal path removed.

**Pull request merged**